nvflare/fuel/f3/drivers/tcp_driver.py

```diff
@@ -0,0 +1,106 @@
+# Copyright (c) 2021-2023, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+import os
+import socket
+from socketserver import TCPServer, ThreadingTCPServer
+from typing import Any, Dict, List
+
+from nvflare.fuel.f3.drivers.base_driver import BaseDriver
+from nvflare.fuel.f3.drivers.driver import ConnectorInfo, Driver
+from nvflare.fuel.f3.drivers.driver_params import DriverCap, DriverParams
+from nvflare.fuel.f3.drivers.net_utils import get_ssl_context, get_tcp_urls
+from nvflare.fuel.f3.drivers.socket_conn import ConnectionHandler, SocketConnection
+
+log = logging.getLogger(__name__)
+
+
+class TcpStreamServer(ThreadingTCPServer):
+
+    TCPServer.allow_reuse_address = True
+
+    def __init__(self, driver: Driver, connector: ConnectorInfo):
+        self.driver = driver
+        self.connector = connector
+
+        params = connector.params
+        self.ssl_context = get_ssl_context(params, ssl_server=True)
+
+        host = params.get(DriverParams.HOST.value)
+        port = int(params.get(DriverParams.PORT.value))
+        self.local_addr = f"{host}:{port}"
+
+        TCPServer.__init__(self, (host, port), ConnectionHandler, False)
+
+        if self.ssl_context:
+            self.socket = self.ssl_context.wrap_socket(self.socket, server_side=True)
+
+        try:
+            self.server_bind()
+            self.server_activate()
+        except BaseException as ex:
+            log.error(f"{os.getpid()}: Error binding to  {host}:{port}: {ex}")
+            self.server_close()
+            raise
+
+
+class TcpDriver(BaseDriver):
+    def __init__(self):
+        super().__init__()
+        self.server = None
+
+    @staticmethod
+    def supported_transports() -> List[str]:
+        return ["tcp", "stcp"]
+
+    @staticmethod
+    def capabilities() -> Dict[str, Any]:
+        return {DriverCap.HEARTBEAT.value: False, DriverCap.SUPPORT_SSL.value: True}
+
+    def listen(self, connector: ConnectorInfo):
+        self.connector = connector
+        self.server = TcpStreamServer(self, connector)
+        self.server.serve_forever()
+
+    def connect(self, connector: ConnectorInfo):
+        self.connector = connector
```

Comment:
Why does `TCPDriver` keeps changing its `self.connector`?
In every listen and connect method.
If we need it, do we need to lock it since these methods can be called in diff. threads?
If we don't need it, should we just remove it?

Reply:
Connector is needed in the driver but it's not available at driver construction time so it's passed in through those 2 methods.

Those methods are only called from one thread so lock is not needed.

Reply:
Will we re-use the driver class?

If we re-used the driver class, it's `self.connector` will keep changing?

Reply:
Yes, the driver can be reused for different connectors if implemented properly but it's not reused in current implementation for simplicity.

Reply:
Got it, we might need to take care of this `self.connector` if we want to reuse this in the future.
For now since we are not reusing the driver, it is OK.